### PR TITLE
Compile and render Synapse's docs into a browsable, mobile-friendly and searchable website

### PR DESCRIPTION
**Pull request by anoadragon453**
**View the demo!** [https://anoadragon453.github.io/synapse/](https://anoadragon453.github.io/synapse/latest/index.html)

Synapse's documentation has traditionally existed as a collection of notes inside our `docs/` directory. While this has the benefit of being close to the code (such that it can be updated alongside code changes), it suffers from a lack of discoverability, searchability and presentation in general.

This PR attempts to organise our existing documentation into a searchable and user-friendly website for the benefit of both users of Synapse (as they can quickly browse and search for questions about running the homeserver), as well as Synapse developers in terms of quickly seeing what components of the system are documented, and which aren't.

For this project I chose [mdbook](https://github.com/rust-lang/mdBook/). Partly because we're [already using it](https://matrix-org.github.io/matrix-appservice-irc/latest/index.html) for other Matrix-related projects, but also partly because of the extensibility of the UI and rendering, quick build times and great UX.

I've done by best to organise the changes into logical commits, with useful commit messages. As a basic overview of how mdbook works however:

* `book.toml` is a new file at the root of the repository that configures how mdbook behaves, the features of the site and where to look for the content.
* `docs/SUMMARY.md` defines the outline of the site (and is what is rendered into the left sidebar). When you add a new page, it needs to be linked to from this file.
* `docs/website_files` contains custom HTML, CSS and JS that's injected into the resulting static site files at build time. These files are defined in `book.toml`, and we utilise this extensibility to add a floating table of contents plugin, and to tweak some of the default UI.

This PR strictly keeps the old documentation where it was so that historical links don't break. That being said, ideally new documentation pages would land in a sensible location rather than the root of the `docs/` directory. I've explained a little bit about the new structure (and what to do when adding a new page) in [docs/README.md](https://github.com/matrix-org/synapse/blob/anoa/mdbook/docs/README.md#adding-to-the-documentation).

All that being said, there's still a few open-ended questions that I'd like to discuss on this pull request:

1. I've configured the CI through Github Actions to push the built documentation to the `gh-pages` branch via Github Actions. This means that the documentation will initially be available at https://matrix-org.github.io/synapse, and we could configure our DNS to point a more official-looking URL at it for the long term.

    The Github Actions configuration (located [here](https://github.com/matrix-org/synapse/blob/anoa/mdbook/.github/workflows/docs.yaml)) is currently configured to push to the `latest` version of the website on every new commit on `develop`.

    This is much easier to maintain than configuring CI to push to matrix.org's infrastructure, but does it sound reasonable to you? Note that mdbook supports versioning, but I've opted to keep that to a separate PR to minimise review scope.

2. I'd like to direct people to the website as much as possible, but appreciate that some people will still be going into the `docs/` directory, either because they're not aware of the new site or just personal preference. mdbook is handy as the files are still just regular old markdown files.

    However, I appreciate that there are now [several new folders](https://github.com/matrix-org/synapse/tree/anoa/mdbook/docs) living in the `docs/` directory, and at present many of them only lead to a couple files (such as [section headers](https://github.com/matrix-org/synapse/blob/anoa/mdbook/docs/usage/configuration/README.md), which don't make much sense if you're not looking at them on the website). I don't think these are much of a nuisance, but makes things slightly more confusing if you're just looking at codebase standalone.
3. I'm not sure what to do about the [privacy policy templates](https://github.com/matrix-org/synapse/tree/anoa/mdbook/docs/privacy_policy_templates/en) in the `docs/` directory.

I'd also like to note that while the current state of the website is quite nice, there's still a few warts:

* Some of our documentation is still in RST. I'd like to update these to be markdown, but doing so would break existing links. The most notable example is UPGRADE.rst - but perhaps we can just link to this file instead of hosting it on the docs site?
* The welcome and overview page is currently quite bare. I'd like to see some links to interesting parts of the docs for different types of visitors - such as developers versus sysadmins.
* ~~mdbook supports having a button on each page that will take you to github to edit the page contents. Unfortunately, there's a bug where this feature does not produce the correct edit link if you've customised your documentation directory path (the default is `src/`, we're using `docs/`). Thus this feature is currently disabled in the config, but we can re-enable it once it's been fixed ([upstream bug](https://github.com/rust-lang/mdBook/issues/1543)).~~ fixed in mdbook 0.4.9!

For this reason I'm not putting a direct link to the documentation at the top of the root README yet, but hope to do so after we get the basic infrastructure in and refine things a little!

Requires: ~~https://github.com/matrix-org/synapse/pull/10089.~~

### CODE REVIEW

File: .github/workflows/docs.yaml
Comment (clokep) [inline .github/workflows/docs.yaml:25]:
```suggestion
      - name: Build the documentation
```

File: .github/workflows/docs.yaml
Comment (clokep) [inline .github/workflows/docs.yaml:4]: Should we remove this for now? 🤷

File: .github/workflows/docs.yaml
Comment (clokep) [inline .github/workflows/docs.yaml:34]: I wonder if we should call this `develop`? I think `latest` usually means "latest stable version"?

File: docs/development/contributing_guide.md
Comment (clokep) [inline docs/development/contributing_guide.md:2]:
```suggestion
  Include the contents of CONTRIBUTING.md from the project root (where GitHub likes it 
```

**Review by clokep — COMMENTED** [review]
> This is much easier to maintain than configuring CI to push to matrix.org's infrastructure, but does it sound reasonable to you?

Seems reasonable to me, yes.

> I'm not sure what to do about the privacy policy templates in the docs/ directory.

Seems like we could add a page describing what you need for a privacy policy and then include them, like we did for the sample config?

A couple of other thoughts:

* I think leaving the docs in place for now is the right thing, but we might want to slowly move them to "better" places. We might break some links, but that's what perma links are for.
* I don't know what to do about things like UPGRADE.rst since that's linked in MANY places, but we could possibly leave it and the link to the website showing the upgrade notes?
* I don't think we should hold this up on it being perfect, it will be easier to iterate on once it exists.

File: .github/workflows/docs.yaml
Comment (callahad) [inline .github/workflows/docs.yaml:29]: We should pin this as it's a third party action and a compromise there could be leveraged to compromise the Synapse repo. See [Security Hardening for GitHub Actions](https://docs.github.com/en/actions/learn-github-actions/security-hardening-for-github-actions#using-third-party-actions)

```suggestion
        uses: peaceiris/actions-gh-pages@068dc23d9710f1ba62e86896f84735d869951305 # v3.8.0
```

File: .github/workflows/docs.yaml
Comment (callahad) [inline .github/workflows/docs.yaml:21]: We should pin this as it's a third party action and a compromise there could be leveraged to compromise the Synapse repo. See [Security Hardening for GitHub Actions](https://docs.github.com/en/actions/learn-github-actions/security-hardening-for-github-actions#using-third-party-actions)

```suggestion
        uses: peaceiris/actions-mdbook@4b5ef36b314c2599664ca107bb8c02412548d79d # v1.1.14
```

File: docs/SUMMARY.md
Comment (babolivier) [reply docs/SUMMARY.md:59]: Also, should we use this as an excuse to move the last few remaining rst docs to markdown?

Comment (anoadragon453) [reply docs/SUMMARY.md:59]: It just tries to parse them as markdown, yeah.

From what I'm hearing it sounds like we're OK to break links to smaller documents given people should be sharing permalinks anyways. I'll just convert these files to markdown.

File: docs/usage/configuration/sample_config.md
Comment (anoadragon453) [reply docs/usage/configuration/sample_config.md:28]: Probably yes. While the table of contents does indicate that the Logging Config section is there, it's immediately obvious. The ToC also doesn't show on mobile.

File: .github/workflows/docs.yaml
Comment (anoadragon453) [reply .github/workflows/docs.yaml:4]: heh, this is one of the things that was fixed in the bit I broke out of this PR. But yes, will remove it here.

File: .github/workflows/docs.yaml
Comment (anoadragon453) [reply .github/workflows/docs.yaml:34]: I suppose that would make it more obvious that it tracks `develop`. Can do!

File: docs/SUMMARY.md
Comment (anoadragon453) [reply docs/SUMMARY.md:59]: This ended up being a fairly sizable diff. I've broken it off into another PR: https://github.com/matrix-org/synapse/pull/10089.

File: docs/usage/configuration/sample_config.md
Comment (anoadragon453) [reply docs/usage/configuration/sample_config.md]: I have split them out into separate pages :)

File: changelog.d/10086.doc
Comment (clokep) [inline changelog.d/10086.doc:1]: Should we put a link here to the rendered version or maybe we should add it to the release notes?

File: docs/SUMMARY.md
Comment (clokep) [inline docs/SUMMARY.md:41]:
```suggestion
      - [Presence Router](presence_router_module.md)
      - [Media Providers]()
      - [Writing a Module]()
```

File: docs/SUMMARY.md
Comment (clokep) [inline docs/SUMMARY.md:63]: Should this be under the logging config above?

File: docs/development/feature_documentation/README.md
Comment (clokep) [inline docs/development/feature_documentation/README.md:1]: It feels weird to call this "Feature Documentation" when it is under "development"... I don't have a better name unfortunately.

File: docs/usage/administration/admin_api/README.md
Comment (clokep) [inline docs/usage/administration/admin_api/README.md:15]: This link doesn't seem to work?

File: docs/usage/configuration/README.md
Comment (clokep) [inline docs/usage/configuration/README.md:4]:
```suggestion
file should have been generated when you [installed Synapse](../../setup/installation.html).
```

**Review by clokep — APPROVED** [review]
Seems like a great first cut to me! I'm sure there's improvements to make but we can probably do those bit by bit.

File: changelog.d/10086.doc
Comment (anoadragon453) [reply changelog.d/10086.doc:1]: Release notes sounds good - we'll need to flip the switch on the GitHub Pages settings once we've merged anyhow.

File: docs/SUMMARY.md
Comment (anoadragon453) [reply docs/SUMMARY.md:41]: Huh, what *is* a media provider?

File: docs/SUMMARY.md
Comment (anoadragon453) [reply docs/SUMMARY.md:63]: Mm, yes probably!

File: docs/development/feature_documentation/README.md
Comment (anoadragon453) [reply docs/development/feature_documentation/README.md:1]: Hmm, "Component Documentation" perhaps?

File: docs/usage/administration/admin_api/README.md
Comment (anoadragon453) [reply docs/usage/administration/admin_api/README.md:15]: Ah yes, we can only direct link to things that will be included in the documentation. This script won't, obviously. This probably doesn't need to be a link ([the actual script itself is tiny](https://github.com/matrix-org/synapse/blob/develop/scripts/register_new_matrix_user)). I've called it a command instead and provided a bit more detail about where one may find it locally.

File: docs/development/feature_documentation/README.md
Comment (clokep) [reply docs/development/feature_documentation/README.md:1]: "Internal Documentation"? "Design Documentation"?

File: docs/SUMMARY.md
Comment (anoadragon453) [reply docs/SUMMARY.md]: #10089 has now merged, and the changes that were based on it have been incorporated in 170fed8e1d.

File: docs/development/feature_documentation/README.md
Comment (anoadragon453) [reply docs/development/feature_documentation/README.md:1]: Internal Documentation sounds good!

File: docs/SUMMARY.md
Comment (clokep) [reply docs/SUMMARY.md:42]: An example is the S3 storage provider: https://github.com/matrix-org/synapse-s3-storage-provider

I think the code tends to call them media providers, but maybe storage provider is a better name?

Comment (anoadragon453) [reply docs/SUMMARY.md:42]: Aha, apparently they're called "media_storage_providers" as the config option. I guess that works...?

Comment (clokep) [reply docs/SUMMARY.md:42]: WFM.

File: docs/README.md
Comment (tulir) [inline docs/README.md:3]:
```suggestion
**The documentation is currently hosted [here](https://matrix-org.github.io/synapse).**
```

File: docs/README.md
Comment (tulir) [inline docs/README.md:11]:
```suggestion
[website](https://matrix-org.github.io/synapse) for easier discoverability.
```

**Pull request merged**
